### ISSUE TITLE
Use type hints on function vars to infer return types

### ISSUE THREAD

**Issue by yuhan0** [clj-kondo/clj-kondo#690]
Type hints on function vars can be used to infer their return types, see 
https://ask.clojure.org/index.php/8993/where-to-place-function-return-type-hints

Currently only tags on the arglists are used for type checking:
```clj
(defn f1 ^String [])
(defn ^String f2 [])

(inc (f1)) ;; <- warning
(inc (f2)) ;; <- no warning
```
**Additional context**

If hints are placed in both positions, the arglist hint takes precedence.
```clj
(defn ^long f3 ^String [])

(inc (f3)) ;; <- should produce a warning
```

Since the arglist hint is the preferred and less error-prone location for return type hints,
clj-kondo could also possibly raise a lint warning when the var is hinted in a `defn`, since it wrongly tags the function objects.

(Might be interested to do a PR myself if I have the time)

**Comment by borkdude** [comment]
PR welcome. Let me know if you start working on this, so we won't duplicate efforts.

**Comment by borkdude** [comment]
Reopen

**Comment by borkdude** [comment]
`(inc (f1))` now warns:

<img width="502" alt="Screenshot 2023-05-03 at 12 21 25" src="https://user-images.githubusercontent.com/284934/235890977-d0ae6684-ff7a-4a18-a2e6-6236d6a640bd.png">

**Comment by borkdude** [comment]
The other one is an incorrectly placed type hint which clj-kondo will warn about. As soon as you fix it, you will get the previous example which already works.
